### PR TITLE
Split Var into TmVar, TmName, TpVar, TpName, Tag

### PR DESCRIPTION
**Pull request by ccshan**
https://github.com/diprism/perpl/issues/103#issuecomment-1207199654

### CODE REVIEW

**Comment by davidweichiang** [comment]
Oops, I left a bunch of comments on the branch instead of on the PR. That's surprising that they are two different things. Sorry about that!

File: src/Struct/Exprs.hs [hunk 1]
```diff
@@ -14,16 +16,16 @@ data UsProgs = UsProgs [UsProg] UsTm    -- definitions, main
...
+  | UsProgExtern TmName Type            -- lhs, type
+  | UsProgData TpName [TpVar] [Ctor]    -- lhs, type params, constructors
   deriving (Eq, Ord)
 
 -- Scheme-ified definition
 data SProg =
-    SProgDefine Var [Var] [Var] Term Type  -- lhs, tags, type params, rhs, type
-  | SProgExtern Var Type                -- lhs, type
-  | SProgData Var [Var] [Var] [Ctor]    -- lhs, tags, type params, constructors
+    SProgDefine TmName [Tag] [TpVar] Term Type  -- lhs, tags, type params, rhs, type
```
Comment (davidweichiang) [inline src/Struct/Exprs.hs:26]: You decided not to separate tags from tag variables?

File: src/Scope/Subst.hs [hunk 2]
```diff
@@ -215,24 +261,14 @@ instance Substitutable a => Substitutable (Maybe a) where
...
 instance Substitutable v => Substitutable (Map.Map k v) where
   substM m = sequence (Map.map substM m)
   freeVars = error "freeVars called on a Subst (undefined behavior)"
 
 instance Substitutable UsTm where
-  substM (UsVar x) =
-    pure UsVar <*> substVar x id (const x) (const x) (const x) x
+  substM (UsVar x) = fmap (f . Map.findWithDefault (Rename x) x . tmVars) get
+    where f (Rename y) = UsVar y
+          f (Replace _) = UsVar x
```
Comment (davidweichiang) [inline src/Scope/Subst.hs:271]: Why can't UsVars be replaced with Terms?

File: src/Scope/Subst.hs [hunk 3]
```diff
@@ -1,92 +1,131 @@
...
+import qualified Scope.Ctxt as C
+import Scope.Fresh (newVar, newVars)
 import Struct.Lib
 
 ----------------------------------------
 
-data SubT = SubVar Var | SubTm Term | SubTp Type | SubTg Tag
-  deriving (Eq, Ord, Show)
-type Subst = Map Var SubT
+data STerm = Rename TmVar -- `Rename v` means `TmVarL v unchanged`
```
Comment (davidweichiang) [inline src/Scope/Subst.hs:23]: Could you clarify this comment a bit? 

What's the difference between `Rename v` and `Replace (TmVarL v)`?

Is it possible you mean "`Rename v` means 'Change `TmVarL`, but not `UsVar`'"?

Although that's what `SubTerm` currently does, I suspect there would be no problem with changing `UsVar` as well. In that case, I suspect that `Rename v` could be eliminated, and `tmVars` could just be a `Map TmVar Term`.

File: src/Scope/Subst.hs [hunk 4]
```diff
@@ -126,11 +165,13 @@ subst s a = runSubst s (substM a)
 
 -- Run substM under a given context
 substWithCtxt :: Substitutable a => Ctxt -> Subst -> a -> a
-substWithCtxt g s = subst (Map.union (Map.mapWithKey (const . SubVar) g) s)
+substWithCtxt g = subst . remember (C.tmVars  g)
```
Comment (davidweichiang) [inline src/Scope/Subst.hs:168]: This means that local variables won't be alpha-renamed apart from either global defines or datatype names. But I guess I don't see why that would be a problem.

**Comment by davidweichiang** [comment]
This does not give an error:
```
data x = Bla;
let x = () in (fail : x)
```
But this does:
```
data Bla = Bla;
()
```

**Comment by davidweichiang** [comment]
Since both Ctxt and Subst have three mappings for types, terms, and tags, I think the easiest way to make things consistent is to let types, terms, and tags have three different namespaces. The above commit makes this change. The rest of my comments are minor.

**Review by davidweichiang — APPROVED** [review]

File: src/Struct/Exprs.hs (hunk 1 above)
Comment (ccshan) [reply src/Struct/Exprs.hs]: Yeah, we can always split them later if needed...

**Pull request merged**